Add disk based edge caching support

poornas opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
This PR adds disk-based edge caching support for minio server.

Cache settings can be configured in config.json to take a list of disk drives,
cache expiry in days and file patterns to exclude from cache or via environment
variables MINIO_CACHE_DRIVES, MINIO_CACHE_EXCLUDE and MINIO_CACHE_EXPIRY

The design assumes that Atime support is enabled and the list of cache drives is
fixed.
 - Objects are cached on both GET and PUT/POST operations.
 - Expiry is used as a hint to evict older entries from the cache, or if 80% of the cache
   capacity is filled.
 - When object storage backend is down, GET, LIST and HEAD operations fetch
   object seamlessly from a cache.

Current Limitations
 - Bucket policies are not cached, so anonymous operations are not supported in
   offline mode.
 - Objects are distributed using deterministic hashing among the list of cache
   drives specified.If one or more drives go offline, or cache 
   drive configuration is altered - performance could degrade to linear lookup.

## Description
<!--- Describe your changes in detail -->

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Manually and with Mint on XL,FS and gateway in online mode.Manually in offline mode.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [x] All new and existing tests passed.

harshavardhana commented on cmd/disk-cache.go:
```diff
@@ -0,0 +1,477 @@
+package cmd
```
license header is must.

harshavardhana commented on cmd/disk-cache.go:
```diff
@@ -0,0 +1,477 @@
+	// creation-time of the cache
+	createTime time.Time
+	// cache database
+	db ObjectLayer
```
we shouldn't call this `cache database`

harshavardhana commented on cmd/disk-cache.go:
```diff
@@ -0,0 +1,477 @@
+// Disk usage is high if disk used is > maxUsage
+func (c diskCache) diskUsageHigh() bool {
+	di, err := disk.GetInfo(c.dir)
+	if err != nil {
```
eating the error is a bad idea.

harshavardhana commented on cmd/disk-cache.go:
```diff
@@ -0,0 +1,477 @@
+func (c diskCache) diskAvailable(size int64) bool {
+	di, err := disk.GetInfo(c.dir)
+	if err != nil {
+		return true
```
eating the error is a bad idea.

harshavardhana commented on cmd/disk-cache.go:
```diff
@@ -0,0 +1,477 @@
+
+// Returns metadata of the cached object
+func (c diskCache) GetObjectInfo(bucket, object string) (info ObjectInfo, err error) {
+	objInfo, err := c.db.GetObjectInfo(bucket, object)
```
one line `return c.db.GetObjectInfo(bucket, object)`

harshavardhana commented on cmd/disk-cache.go:
```diff
@@ -0,0 +1,477 @@
+
+// Inits the cache directory if it is not init'ed already.
+// Initializing implies creation of new FS Object layer and entries directory that holds atime of cached entries
+func newDiskCache(dir string, maxUsage int) (*diskCache, error) {
```
Move newThis() to a starting of the function and then implement the necessary calls.

harshavardhana commented on cmd/disk-cache.go:
```diff
@@ -0,0 +1,477 @@
+		return nil, err
+	}
+
+	cache := &diskCache{dir, maxUsage, time.Now().UTC(), fs, make(chan struct{})}
```
time.Now().UTC() can use `UTCNow()`

harshavardhana commented on cmd/disk-cache_test.go:
```diff
@@ -0,0 +1,87 @@
+/*
+ * Minio Cloud Storage, (C) 2016 Minio, Inc.
```
should be 2017

harshavardhana commented on cmd/object-api-datatypes.go:
```diff
@@ -115,6 +115,9 @@ type ObjectInfo struct {
 	HealObjectInfo *HealObjectInfo `xml:"HealObjectInfo,omitempty"`
+
+	// Date and time when the object was last accessed
+	ATime time.Time
```
Call it AccTime

harshavardhana commented on cmd/object-api-datatypes.go:
```diff
@@ -115,6 +115,9 @@ type ObjectInfo struct {
 	UserDefined    map[string]string
 	HealObjectInfo *HealObjectInfo `xml:"HealObjectInfo,omitempty"`
+
+	// Date and time when the object was last accessed
```
missing . at the end.

harshavardhana commented on cmd/utils.go:
```diff
@@ -229,3 +231,44 @@ func toS3ETag(etag string) string {
+
+// GetLastAccess gets last accessed date and time of the object.
+func GetLastAccess(int interface{}) (aTime time.Time) {
+	stat, ok := int.(*syscall.Stat_t)
```
We should make this multi platform, this cannot compile properly and work on windows.  Perhaps create a package like this https://github.com/djherbis/atime

ngbrown replied:
Was this resolved? I tried to enable caching on Windows and got the error: `ERROR Unable to initialize disk caching: Atime support required for disk caching`. Last access time is enabled on the drive:

```
D:\>fsutil behavior query DisableLastAccess
DisableLastAccess = 0
```

harshavardhana replied:
@ngbrown caching is only supported on Linux currently, use it accordingly.

ngbrown replied:
@harshavardhana Ok. I'm forced to have the primary data drive be a high capacity file share from a network appliance.  I wanted to cache on the local, fast flash drive to reduce network traffic.  If I just have to make an `atime` compatible package for Windows, I can look into that.

harshavardhana replied:
By all means, all you have to fix is time package which is use to support windows.

krishnasrinivas reviewed (dismissed):
As discussed `for k, v := range c.policyDiskMap` can range in random order, hence it should be kept in slice instead of map.

Also atime support needs to be checked during startup.

what can be done is:
write to a tmp file and note the atime
sleep for 10ms
read from the tmp file and see if atime changes

if atime does not change then error out.

harshavardhana commented on cmd/api-errors.go:
```diff
@@ -163,6 +163,8 @@ const (
 	ErrPartsSizeUnequal
 	ErrInvalidRequest
+	ErrBackendDown
+	ErrATimeUnsupported
```
Where is this Error being used? if we are validating if backend supports atime in startup we don't need this.

harshavardhana commented on cmd/disk-cache.go:
```diff
@@ -0,0 +1,490 @@
+	delimiter := "/"
+	maxKeys := 1000
+	for {
+		for {
```
```
for !c.diskUsageLow() {
}
``` 

This is what you want really?

harshavardhana commented on cmd/disk-cache.go:
```diff
@@ -0,0 +1,490 @@
+				break
+			}
+			// expire objects older than one month since accessed
+			expiry := time.Now().AddDate(0, 0, -30)
```
Perhaps use UTCNow().AddDate?

harshavardhana commented on cmd/disk-cache.go:
```diff
@@ -0,0 +1,490 @@
+							continue
+						}
+						if err = c.db.DeleteObject(bucket.Name, object.Name); err != nil {
+							errorIf(err, "Unable to remove cache entry in dir %s/%s", bucket.Name, object.Name)
```
Even if there is an error deletedCount is calculating ++ we can avoid that?

poornas replied:
fixed this.

harshavardhana commented on cmd/disk-cache.go:
```diff
@@ -0,0 +1,490 @@
+			break
+		}
+		if n > 0 {
+			pipeWriter.Write(buf)
```
Can we not simplify this with using io.MultiWriter(pipeWriter, wPipe) and then do `io.Copy(w, r)` ?

harshavardhana commented on cmd/fs-v1.go:
```diff
@@ -632,8 +636,9 @@ func (fs fsObjects) PutObject(bucket string, object string, data *hash.Reader, m
 	}
-
-	metadata["etag"] = hex.EncodeToString(data.MD5Current())
+	if metadata["etag"] == "" {
```
This is not needed, "etag" will always be the MD5Current()

poornas replied:
for large uploads, the ETag generated during multipart upload to backend will not match the ETag generated using single Put operation for caching to FS. Hence the cache Put operation needs to pass the Etag in the metadata headers so that the backend's ETag can be used on the cached object as well. This check is required unless caching also uses multi-part put operation.

harshavardhana replied:
This looks like design bug. It's changing code to cater  for caching requirement .. such type of spillage shouldn't happen.

harshavardhana commented on cmd/gateway-handlers.go:
```diff
@@ -918,7 +923,9 @@ func (api gatewayAPIHandlers) GetBucketLocationHandler(w http.ResponseWriter, r
 
-	if _, err := getBucketInfo(bucket); err != nil {
+	_, err := getBucketInfo(bucket)
+	_, backendDown := errorCause(err).(BackendDown)
```
how about doing a switch case? 
```
switch errorCause(err).(type) {
case BackendDown:
default:
```

poornas replied:
switch will not be very useful here - both backendDown and err == nil needs to be handled identically.

harshavardhana commented on cmd/gateway-router.go:
```diff
@@ -52,11 +52,19 @@ type gatewayAPIHandlers struct {
 		ObjectAPI: func() GatewayLayer { return gw },
 		objectAPIHandlers: objectAPIHandlers{
 			ObjectAPI: newObjectLayerFn,
+			cache:     cache,
```
I would suggest that we make cache like ObjectAPI, which can be re-initialized from admin API. Such that this element is public. 

```
ObjectAPI: newObjectLayerFn,
Cache:  newCacheFn,
```

krishnasrinivas commented on cmd/config-versions.go:
```diff
@@ -564,3 +565,23 @@ type serverConfigV22 struct {
+	StorageClass storageClassConfig `json:"storageclass"`
+
+	// Cache configuration
+	Cache *cacheConfig `json:"cache"`
```
Correct way to define cache config is:
```
type serverConfigV23 struct {
	sync.RWMutex
	Version string `json:"version"`

	// S3 API configuration.
	Credential auth.Credentials `json:"credential"`
	Region     string           `json:"region"`
	Browser    BrowserFlag      `json:"browser"`
	Domain     string           `json:"domain"`

	// Storage class configuration
	StorageClass storageClassConfig `json:"storageclass"`

	// Cache configuration
	Cache struct {
		Drives  []string `json:"drives"`
		Expiry  int      `json:"expiry"`
		Exclude []string `json:"exclude"`
	} `json:"cache"`

	// Notification queue configuration.
	Notify *notifier `json:"notify"`
}
```

Because for arguments sake let's say in future we decide to change `Expiry` to string type which takes `"30d"` `"12h"`, if we do that json.Unmarshal() will fail during migration. Another way to fix it is do `type CacheConfigV1 struct {}` and include it in the serverConfig23. It is easier to "inline" the struct (as shown above) instead of defining a new struct type.

This argument applies to notifier too. If we change the type of any of the fields then migration will break.

krishnasrinivas commented on cmd/disk-cache-config.go:
```diff
@@ -0,0 +1,120 @@
+	}
+	return
+}
+
```
Looks like the methods on cacheConfig are not used. So you can remove them. You can remove cacheConfig too I think?

krishnasrinivas commented on cmd/disk-cache.go:
```diff
@@ -0,0 +1,764 @@
+}
+
+// Return error if Atime is disabled on the O/S
+func checkAtimeSupport() (err error) {
```
what needs to be done is, the file needs to be created on the cache-dirs, using os.TempDir() will use `/tmp`

Also if a partition is mounted with atime disabled, `os.Chtimes(file.Name(), atime, mtime)` will still succeed.

So, once you create a file, note the atime using os.Stat(), then perform a Read() operation and do os.Stat() again to see if the Atime changed.

krishnasrinivas commented on cmd/disk-cache-format.go:
```diff
@@ -0,0 +1,440 @@
+/*
```
you can rename this file to format-disk-cache.go similar to format-fs.go and format-xl.go

krishnasrinivas commented on cmd/disk-cache-format.go:
```diff
@@ -0,0 +1,440 @@
+	// Format config file carries cache format specific details.
+	cacheFormatConfigFile = "format.json"
+
+	formatMetaCacheVersionV1 = "1"
```
This should be removed and formatMetaVersionV1 should be used instead.

krishnasrinivas commented on cmd/disk-cache-format.go:
```diff
@@ -0,0 +1,440 @@
+	formatCacheVersionV1 = "1"
+
+	// Format config file carries cache format specific details.
+	cacheFormatConfigFile = "format.json"
```
You can re-use `formatConfigFile` as we are consistently using the file name format.json for xl and fs.

krishnasrinivas commented on cmd/disk-cache-format.go:
```diff
@@ -0,0 +1,440 @@
+	formatMetaV1
+	Cache struct {
+		Version string `json:"version"` // Version of 'cache' format.
+		Dir     string `json:"dir"`
```
What we can do is remove `Dir` field and use `const cacheDataDir = data` like how we discussed

krishnasrinivas commented on cmd/disk-cache.go:
```diff
@@ -0,0 +1,764 @@
+		return putObjectPartFn(bucket, object, uploadID, partID, data)
+	}
+	//Hold the lock so that disk cache and fs put operations for each part get locks in same order.
+	objectMPartPathLock := dcache.nsMutex.NewNSLock(pathJoin(cacheFSObjectsLockPrefix, minioMetaMultipartBucket), pathJoin(bucket, object))
```
you need not hold the lock here as cache will not be shared with other minio instances.

krishnasrinivas commented on cmd/disk-cache.go:
```diff
@@ -0,0 +1,764 @@
+const (
+	// disk cache needs to have cacheSizeMultiplier * object size space free for a cache entry to be created.
+	cacheSizeMultiplier      = 100
+	maxDiskUsagePct          = 80 // in %
```
you can call this `cacheMaxDiskUsage` so that all cache global vars start with `cache`

krishnasrinivas commented on cmd/disk-cache.go:
```diff
@@ -0,0 +1,764 @@
+	h := fnv.New32a()
+	h.Write([]byte(s))
+	return int(h.Sum32()) % len(c.cfs)
+}
```
you can use the same function used by large bucket:
```
func crcHashMod(bucket, object string) int {
        return crc32.Checksum([]byte(key), crc32.IEEETable) % len(c.cfs)
}
```

krishnasrinivas commented on cmd/disk-cache.go:
```diff
@@ -0,0 +1,764 @@
+	var matchDisk *cacheFSObjects
+	firstMatchDisk := false
+	for {
+		if c.cfs[i].getOnline() {
```
you can rename `getOnline()` to `IsOnline()` (we follow this convention)

krishnasrinivas commented on cmd/disk-cache.go:
```diff
@@ -0,0 +1,764 @@
+		return matchDisk, nil
+	}
+	return nil, errDiskNotFound
+}
```
If you break the above function into two functions, its more readable and easy on the eyes:
```
func (c diskCache) getCacheFSRead(bucket, object string) (*cacheFSObjects, error) {
        index := c.hashIndex(bucket, object)
        numDisks := len(c.cfs)
        i := index
        var firstOnlineDisk *cacheObjects
        for {
                if c.cfs[i].IsOnline() {
                        if firstOnlineDisk == nil {
                                firstOnlineDisk = c.cfs[i]
                        }
                        if c.cfs[i].Exists(bukcet, object) {
                                return c.cfs[i], nil
                        }
                }
                i = (i + 1) % numDisks
                if i == index {
                        break
                }
        }
        if firstOnlineDisk != nil {
                return firstOnlineDisk, nil
        }
        return nil, errDiskNotFound
}

func (c diskCache) getCacheFSWrite(bucket, object string) (*cacheFSObjects, error) {
        index := c.hashIndex(bucket, object)
        numDisks := len(c.cfs)
        i := index

        for {
                if c.cfs[i].IsOnline() {
                        return c.cfs[i], nil
                }
                i = (i + 1) % numDisks
                if i == index {
                        break
                }
        }
        return nil, errDiskNotFound
}
```

harshavardhana commented on .gitignore:
```diff
@@ -24,3 +24,4 @@ prime/
 snap/.snapcraft/
 stage/
 .sia_temp/
+buildcoveragecoverage.txt
```
We don't need this file anymore.

harshavardhana commented on cmd/config-current.go:
```diff
@@ -185,6 +204,9 @@ func newServerConfig() *serverConfig {
 	srvCfg.Notify.Webhook = make(map[string]webhookNotify)
 	srvCfg.Notify.Webhook["1"] = webhookNotify{}
 
+	srvCfg.Cache.Drives = make([]string, 0)
```
This is not needed..

harshavardhana commented on cmd/config-current_test.go:
```diff
@@ -345,56 +345,56 @@ func TestConfigDiff(t *testing.T) {
 		{
-			&serverConfig{Notify: notifier{AMQP: map[string]amqpNotify{"1": {Enable: true}}}},
-			&serverConfig{Notify: notifier{AMQP: map[string]amqpNotify{"1": {Enable: false}}}},
+			&serverConfig{Notify: &notifier{AMQP: map[string]amqpNotify{"1": {Enable: true}}}},
```
any reasons why this PR is concerned with making 'notifier'  as '&notifier' ?

poornas replied:
fixed this - due to rebasing

harshavardhana commented on cmd/config-migrate.go:
```diff
@@ -1807,3 +1813,115 @@ func migrateV21ToV22() error {
+	srvConfig.Domain = cv22.Domain
+
+	// Init cache config.For future migration, Cache config needs to be copied over from previous version.
+	srvConfig.Cache.Drives = make([]string, 0)
```
no need to do this.. for string slices.

poornas replied:
If an empty map is not created, the json converter saves the entry in config.json as null instead of an empty slice []. Hence the make is required.

harshavardhana replied:
if that is the case we should use the `omitempty` params for json tags, such that `null` is never written to disk.

harshavardhana commented on cmd/disk-cache-config.go:
```diff
@@ -0,0 +1,58 @@
+/*
+ * Minio Cloud Storage, (C) 2017 Minio, Inc.
```
license can point to 2018 now.

harshavardhana commented on cmd/disk-cache.go:
```diff
@@ -0,0 +1,894 @@
+/*
+ * Minio Cloud Storage, (C) 2017 Minio, Inc.
```
license can be 2018

harshavardhana commented on cmd/disk-cache.go:
```diff
@@ -0,0 +1,894 @@
+}
+
+// CacheObjectLayer implements primitives for cache object API layer.
+type CacheObjectLayer interface {
```
why do we need a separate interface? we could simply leverage existing object layer interface.. and rest of the stub functions can have overrides.

harshavardhana commented on cmd/disk-cache.go:
```diff
@@ -0,0 +1,894 @@
+// Uses cached-object to serve the request. If object is not cached it serves the request from the backend and also
+// stores it in the cache for serving subsequent requests.
+func (c cacheObjects) GetObject(bucket, object string, startOffset int64, length int64, writer io.Writer, etag string) (err error) {
+	GetObjectFn := c.GetObjectFn
```
are these variables necessary? can we not simply do `c.GetObjectFn` everywhere?

harshavardhana commented on cmd/disk-cache.go:
```diff
@@ -0,0 +1,894 @@
+
+// initCaches initializes the cacheFSObjects for the "drives" specified in config.json
+// or the global env overrides.
+func initCaches(c CacheConfig) (*diskCache, error) {
```
we should call this `newCache` - `init` is not meaningful since this returns values..

harshavardhana replied:
the return value should comply with an interface here..

harshavardhana commented on cmd/fs-v1-multipart.go:
```diff
@@ -374,6 +374,7 @@ func (fs *fsObjects) ListObjectParts(bucket, object, uploadID string, partNumber
 	}
 
 	uploadIDDir := fs.getUploadIDDir(bucket, object, uploadID)
+
```
Avoid adding spaces in this PR, there are no caching related changes in this file.

poornas replied:
fixed.

donatello replied:
The unrelated spaces still seem to be there in the latest version.

harshavardhana commented on cmd/fs-v1.go:
```diff
@@ -98,6 +99,22 @@ func newFSObjectLayer(fsPath string) (ObjectLayer, error) {
 		return nil, errInvalidArgument
 	}
+	fs, err := newFSObjects(fsPath)
+	fatalIf(err, "Unable to initialize FS object layer")
```
We shouldn't fail here when called inside a function which returns error..

poornas replied:
ok, returning error instead.

harshavardhana commented on cmd/fs-v1.go:
```diff
@@ -666,8 +676,9 @@ func (fs *fsObjects) PutObject(bucket string, object string, data *hash.Reader,
 	// No metadata is set, allocate a new one.
-	if metadata == nil {
-		metadata = make(map[string]string)
+	meta := make(map[string]string)
```
what is the need for this copy?

poornas replied:
To avoid concurrent map read and write panic  - fs putObject modifies the metadata map by inserting ETag into the map while cache fs ranges over the k-v in the metadata and copies the entries.

harshavardhana commented on cmd/routers.go:
```diff
@@ -29,6 +29,11 @@ func newObjectLayerFn() (layer ObjectLayer) {
 
+func newCacheObjectsFn() (cacheLayer CacheObjectLayer) {
+	cacheLayer = globalCacheObjectAPI
+	return
```
any reason for this return? -

poornas replied:
removed this.

harshavardhana commented on docs/disk-caching/README.md:
````diff
@@ -0,0 +1,53 @@
+wildcard patterns to exclude certain content from cache as 
+configuration settings. 
+```
+"cache": {
````
The indentation is off here for json can you fix?

poornas replied:
done.

harshavardhana reviewed (changes requested):
Can you also fix the CI errors? @poornas

donatello commented on cmd/api-errors.go:
```diff
@@ -824,6 +823,11 @@ var errorCodeResponse = map[APIErrorCode]APIError{
 	},
+	ErrBackendDown: {
+		Code:           "XMinioBackendDown",
+		Description:    "Network error. Check your network connection",
```
The backend may be unreachable for reasons other than network connectivity - it is better to state only what is known "Object storage backend is unreachable".

donatello commented on cmd/test-utils_test.go:
```diff
@@ -570,6 +578,10 @@ func resetTestGlobals() {
 	// Reset global heal state
 	resetGlobalHealState()
+	//Reset global disk cache flags
+	resetGLobalCacheEnvs()
```
`L` need not be upper case here.

donatello commented on cmd/object-handlers.go:
```diff
@@ -107,7 +107,12 @@ func (api objectAPIHandlers) GetObjectHandler(w http.ResponseWriter, r *http.Req
-	objInfo, err := objectAPI.GetObjectInfo(bucket, object)
+	getObjectInfo := objectAPI.GetObjectInfo
+	if api.CacheAPI() != nil {
+		getObjectInfo = api.CacheAPI().GetObjectInfo
```
Since `globalCacheObjectAPI` is mutated only at server start (in API functions registration), perhaps there could be a way to set the `getObjectInfo` function in a way that we don't need to check `api.CacheAPI() != nil` on each request?

krishnasrinivas commented on cmd/disk-cache.go:
```diff
@@ -0,0 +1,893 @@
+		return errors.New("Atime not supported")
+	}
+	return
+}
```
As discussed, the following is better:
```
// Return error if Atime is disabled on the O/S
func checkAtimeSupport(dir string) (err error) {
        file, err := ioutil.TempFile(dir, "prefix")
        if err != nil {
                return
        }
        fileName := file.Name()
        defer file.Close()
        defer os.Remove(fileName)

        finfo1, err := os.Stat(file.Name())
        if err != nil {
                return
        }
        if _, err = io.Copy(ioutil.Discard, file); err != io.EOF {
                return
        }
        finfo2, err := os.Stat(file.Name())
        if err != nil {
                return
        }
        if atime.Get(finfo2).Equal(atime.Get(finfo1)) {
                return errors.New("Atime not supported")
        }
        return nil
}
```

krishnasrinivas commented on cmd/format-disk-cache.go:
```diff
@@ -0,0 +1,287 @@
+		// time when fresh disks were supplied.
+		Disks []string `json:"disks"`
+	} `json:"cache"` // Cache field holds cache format.
+}
```
@poornas as discussed you can remove references to any structs that is not related for format (i.e cacheFSObjects)

poornas replied:
done

donatello commented on cmd/disk-cache-fs.go:
```diff
@@ -0,0 +1,424 @@
+func (cfs *cacheFSObjects) NewMultipartUpload(bucket, object string, meta map[string]string, uploadID string) (string, error) {
+	if cfs.diskUsageHigh() {
+		select {
+		case cfs.purgeChan <- struct{}{}:
```
~Would this block a new upload from continuing until some disk space is cleared? Is it expected?~

donatello replied:
Nvm - my bad above!

donatello commented on cmd/disk-cache.go:
```diff
@@ -0,0 +1,888 @@
+		if i == index {
+			break
+		}
+	}
```
Better to replace lines 135-144 with:
```
for k := 0; k < numDisks; k++ {
    i := (index + k) % numDisks
    if c.cfs[i].IsOnline() {
        return c.cfs[i], nil
    }
}
```

poornas replied:
done

donatello commented on docs/disk-caching/README.md:
```diff
@@ -0,0 +1,53 @@
+Disk caching can be turned on by updating the "cache" config
+settings for minio server. By default, this is at `${HOME}/.minio`.
+
+"cache" takes the drives location, duration to expiry and any
```
We should mention the units being used for expiry duration.

harshavardhana reviewed (commented):
After i compiled this change and started testing. I see that ENVs are not updated.

harshavardhana reviewed (changes requested):
Looks like ENVs are not updated for `minio server` or `minio gateway` help text?.

Can we have a common text for both server and gateways to explain about ENVs

```
....
....

ENVIRONMENT VARIABLES:
  ACCESS:
     MINIO_ACCESS_KEY: Custom username or access key of minimum 3 characters in length.
     MINIO_SECRET_KEY: Custom password or secret key of minimum 8 characters in length.

  BROWSER:
     MINIO_BROWSER: To disable web browser access, set this value to "off".

  REGION:
     MINIO_REGION: To set custom region. By default it is "us-east-1".

  UPDATE:
     MINIO_UPDATE: To turn off in-place upgrades, set this value to "off".
....
....
```

harshavardhana reviewed (changes requested):
Start minio server with cached drives

```
MINIO_CACHE_DRIVES=/tmp/d1 minio server /tmp/{1..4}
Configuration file /home/harsha/.minio/config.json migrated from version '22' to '23' successfully.
```

But i don't see the cache drives to be part of config.json

```
jq .cache ~/.minio/config.json
{
  "drives": [],
  "expiry": 90,
  "exclude": []
}
```

We should always save the ENV to config if its not set i.e basically its empty.  This also means that the cache never used by my cached drive /tmp/d1/ - i don't see a `format.json` being written.

--------------------------------------------------------------

When minio starts we should also perhaps show `Cache Capacity` and also enhance `mc admin info` to provide this information as well. 

```
minio server /tmp/{1..4}

Drive Capacity: 47 GiB Free, 104 GiB Total
....
....
```

Current `mc admin` doesn't talk about cache at all.
```
mc admin --insecure info myminio-ssl/
●  localhost:9000
   Uptime : online since 1 minute ago
  Version : 2018-02-28T19:59:20Z
   Region : 
 SQS ARNs : arn:minio:sqs::test2:redis arn:minio:sqs::test1:redis 
    Stats : Incoming 1.1KiB, Outgoing 1.3KiB
  Storage : Total 104GiB, Free 47GiB
    Disks : 4, 0
```

--------------------------------------------------------------
After configuring cached directory at `/tmp/d1` by editing `config.json` , i see some odd directory structure. 
```
ls /tmp/d1
[2018-02-28 12:16:20 PST] 4.0KiB data/
ls /tmp/d1/data/
[2018-02-28 12:16:20 PST] 4.0KiB .minio.sys/
ls /tmp/d1/data/.minio.sys/
[2018-02-28 12:16:20 PST]    50B format.json
[2018-02-28 12:16:20 PST] 4.0KiB multipart/
[2018-02-28 12:16:20 PST] 4.0KiB tmp/
```

why do we need another `data/` prefix? since cached directory is entire directory here. Now if i restart the server i see that new file gets written at `/tmp/d1/format.json` is this expected? should it not be created first? 

```
jq . /tmp/d1/format.json 
{
  "version": "1",
  "format": "cache",
  "cache": {
    "version": "1",
    "this": "420ef41c-0ed0-4ce2-a7b8-be8987f98b8d",
    "disks": [
      "420ef41c-0ed0-4ce2-a7b8-be8987f98b8d"
    ]
  }
}
```

I also ran a test program to cache a 1GiB object by simply doing a GET on it, i don't see it is working right now, since i don't see any cached object in /tmp/d1/data

```
find /tmp/d1/
/tmp/d1/data/.minio.sys/format.json
/tmp/d1/format.json
```

krishnasrinivas commented on .gitignore:
```diff
@@ -24,4 +24,3 @@ prime/
 snap/.snapcraft/
 stage/
 .sia_temp/
-buildcoveragecoverage.txt
```
@poornas did you remove this on purpose?

poornas replied:
@harshavardhana had a review comment "We don't need this file anymore." , so yes - it was removed on purpose.

donatello reviewed (changes requested):
What is the expected behaviour when a cache disk is found to be corrupted (i.e. format.json is missing but the directory is *non* empty)?

When i tried to the server in such a situation I got:

```


Trace: 1: cmd/disk-cache.go:701:cmd.newCache()
       2: cmd/disk-cache.go:750:cmd.newServerCacheObjects()
       3: cmd/api-router.go:37:cmd.registerAPIRouter()
       4: cmd/routers.go:88:cmd.configureServerHandler()
       5: cmd/server-main.go:234:cmd.serverMain()
       6: vendor/github.com/minio/cli/app.go:499:cli.HandleAction()
       7: vendor/github.com/minio/cli/command.go:214:cli.Command.Run()
       8: vendor/github.com/minio/cli/app.go:260:cli.(*App).Run()
       9: cmd/main.go:155:cmd.Main()
      10: main.go:71:main.main()
[2018-03-03T00:05:56.741378534Z] [ERROR] Cache drives validation error (open /mnt/sdl1/aditya/c1/format.json: no such file or directory)
```

However, the server continues to start up.

harshavardhana commented on cmd/gateway/azure/gateway-azure.go:
```diff
@@ -72,6 +72,11 @@ ENVIRONMENT VARIABLES:
      MINIO_BROWSER: To disable web browser access, set this value to "off".
 
+	CACHE:
+		 MINIO_CACHE_DRIVES: List of cache drives delimited by ";"
```
this perhaps whole tabbing is off here you should align this with other entries here...

harshavardhana commented on cmd/object-handlers.go:
```diff
@@ -1158,7 +1179,11 @@ func (api objectAPIHandlers) PutObjectPartHandler(w http.ResponseWriter, r *http
 
-	partInfo, err := objectAPI.PutObjectPart(ctx, bucket, object, uploadID, partID, hashReader)
+	putObjectPart := objectAPI.PutObjectPart
+	if api.CacheAPI() != nil {
```
don't cache if SSE header is set as well, we support multipart SSE-C as well remember.

harshavardhana commented on cmd/object-handlers.go:
```diff
@@ -319,7 +334,6 @@ func (api objectAPIHandlers) CopyObjectHandler(w http.ResponseWriter, r *http.Re
 	vars := mux.Vars(r)
 	dstBucket := vars["bucket"]
```
CopyObject should also disable caching when hasSSECustomerHeader is set. same as CopyObjectPart()

harshavardhana commented on docs/config/README.md:
```diff
@@ -92,6 +92,12 @@ minio server /data
+|:---|:---|:---|
+|``drives``| _[]string_ | List of drives|
+|``exclude`` | _[]string_ | List of wildcard patterns for prefixes to exclude from cache |
+|``expiry`` | _int_ | Days to cache expiry |
```
one space after this before the new heading.

harshavardhana commented on docs/config/config.sample.json:
```diff
@@ -11,6 +11,11 @@
         "rrs": ""
     },
+    "cache": {
+        "Drives": [],
```
are you sure our entries are upper cased?

harshavardhana commented on docs/config/config.sample.json:
```diff
@@ -115,4 +120,4 @@
         }
     }
-}
+}
```
missing newline.

harshavardhana commented on docs/disk-caching/README.md:
```diff
@@ -0,0 +1,53 @@
+## Disk based caching support
```
we can perhaps remove the 'support' word.

harshavardhana commented on ~/cache/d/data/.minio.sys/format.json:
```diff
@@ -0,0 +1 @@
+{"version":"1","format":"fs","fs":{"version":"2"}}
```
this shouldn't be part of the commit.

harshavardhana commented on ~/cache/d/format.json:
```diff
@@ -0,0 +1 @@
+{"version":"1","format":"cache","cache":{"version":"1","this":"b6e6a835-0748-47de-86ac-4f1491b2e186","disks":["b6e6a835-0748-47de-86ac-4f1491b2e186"]}}
```
this shouldn't be part of the commit.

harshavardhana reviewed (changes requested):
There are some changes needed, reviewed it once more.

poornas committed:
Add disk based edge caching support.

This PR adds disk based edge caching support for minio server.

Cache settings can be configured in config.json to take list of disk drives,
cache expiry in days and file patterns to exclude from cache or via environment
variables MINIO_CACHE_DRIVES, MINIO_CACHE_EXCLUDE and MINIO_CACHE_EXPIRY

Design assumes that Atime support is enabled and the list of cache drives is
fixed.
 - Objects are cached on both GET and PUT/POST operations.
 - Expiry is used as hint to evict older entries from cache, or if 80% of cache
   capacity is filled.
 - When object storage backend is down, GET, LIST and HEAD operations fetch
   object seamlessly from cache.

Current Limitations
 - Bucket policies are not cached, so anonymous operations are not supported in
   offline mode.
 - Objects are distributed using deterministic hashing among list of cache
   drives specified.If one or more drives go offline, or cache drive
   configuration is altered - performance could degrade to linear lookup.

Fixes #4026

harshavardhana commented on cmd/fs-v1.go:
```diff
@@ -42,7 +42,8 @@ import (
 	// Path to be exported over S3 API.
 	fsPath string
-
+	// meta json file
```
Need more explanation here why this file name is made configurable.

poornas replied:
done.

harshavardhana commented on cmd/fs-v1.go:
```diff
@@ -181,6 +183,11 @@ func NewFSObjectLayer(fsPath string) (ObjectLayer, error) {
 }
 
+// gets name of meta json file
+func (fs *FSObjects) MetaJSONFile() string {
```
This is a superfluous function it's really not needed since we are simply accessing fs.metaJSONFile

poornas replied:
removed this function.

harshavardhana commented on cmd/disk-cache-fs.go:
```diff
@@ -54,28 +59,22 @@ type cacheFSObjects struct {
 	}
 	var cacheFS cacheFSObjects
 	fsObjects := obj.(*FSObjects)
+	fsObjects.metaJSONFile = cacheMetaJSONFile
```
We should change the `.metaJSONFile` such that we take this an input for the NewFSObjects() and avoid doing this kind of change.

poornas replied:
used your suggestion.

harshavardhana reviewed (dismissed):
LGTM

harshavardhana reviewed (approved):
LGTM
